cmd/geth/config.go

```diff
@@ -294,7 +294,105 @@ func makeFullNode(ctx *cli.Context) (*node.Node, ethapi.Backend) {
 		// than startup without full information and later freeze at hVM Phase 0 activation time.
+
 		for {
+
+			// wait until we have a block header that is above hvm genesis height, then confirm we
+			// have hvm genesis block at the expected height
+			for {
+				select {
+				case <-time.After(1 * time.Second):
+					bh, _, err := vm.TBCFullNode.BlockHeaderBest(ctx.Context)
+					if err != nil {
+						log.Crit(fmt.Sprintf("error getting block header best: %s", err))
+					}
+
+					log.Info(fmt.Sprintf("the current best block height is %d", bh))
+
+					if bh < genesisHeight {
+						continue
+					}
+
+					ghh := genesisHeader.BlockHash()
+					// we're above genesis height, but are we on the best chain with genesis?
+					_, height, err := vm.TBCFullNode.BlockHeaderByHash(ctx.Context, &ghh)
+					if err != nil {
+						log.Crit(fmt.Sprintf("could not get block by hash: %s", err))
+					}
+
+					if height != genesisHeight {
+						log.Crit(fmt.Sprintf("mismatched genesis heights %d != %d", height, genesisHeight))
+					}
+
+					break
+				case <-ctx.Context.Done():
+					return nil, nil
+				}
+				break
+			}
+
+			log.Info("done getting hvm genesis header, waiting until we have all of the blocks")
+
+			// grab the current best block, we will then wait until we have all full blocks up to
+			// this one
+			height, blockHeaderBest, err := vm.TBCFullNode.BlockHeaderBest(ctx.Context)
+			if err != nil {
+				log.Crit(fmt.Sprintf("error getting block header best: %s", err))
+			}
+
+			for {
+				// check if we have the full best block available
+				bhh := blockHeaderBest.BlockHash()
+				available, err := vm.TBCFullNode.FullBlockAvailable(ctx.Context, &bhh)
+				if err != nil {
+					log.Crit(fmt.Sprintf("could not determine if full block available: %s", err))
+				}
+
+				// if it's not available, wait until it is
+				if !available {
+					log.Info(fmt.Sprintf("block not available, will wait: %s:%d", blockHeaderBest.BlockHash().String(), height))
+					time.Sleep(5 * time.Second)
+					continue
+				}
+
+				log.Info("block is available %s:%d", blockHeaderBest.BlockHash().String(), height)
+
+				log.Info("done downloading blocks, now will sync indexers to best")
+
+				// wait until we have fixed all of the missing blocks
+				for {
+					missing, err := vm.TBCFullNode.BlocksMissing(ctx.Context, 100)
+					if err != nil {
+						log.Crit(fmt.Sprintf("could not get blocks missing: %s", err))
+					}
+
+					if len(missing) == 0 {
+						break
+					}
+
+					// for each missing block, try to redownload
+					for _, m := range missing {
+						log.Info("found missing block, will atttempt download: %s:%d", m.Hash, m.Height)
+						_, err := vm.TBCFullNode.DownloadBlockFromRandomPeers(ctx.Context, m.Hash, 1)
+						if err != nil {
+							log.Crit(fmt.Sprintf("could not download block from random peer: %s", err))
+						}
+					}
+
+					// give some time for the redownloading to occur
+					time.Sleep(5 * time.Second)
```

Comment:
Is 5s a good sleep time here?  Is there any way we can trigger this without an explicit sleep?

Reply:
hmm so a few lines before this `DownloadBlockFromRandomPeers` gets called, and I believe that downloads the block asynchronously, so ideally we want to give it a little time to receive a response and process it before checking for the missing block again.